### ISSUE TITLE
feat(provisioner/terraform): allow passing custom environment variable to Terraform binary

### ISSUE THREAD

**Issue by johnstcn** [coder/coder#14452]
In order to support some use-cases around caching Terraform modules and provisioners, we need to allow setting some environment variables _specifically_ for certain Terraform invocations.

One such example would be setting `HTTP_PROXY` / `HTTPS_PROXY` to a caching proxy during `terraform init`, so that cached versions of providers and modules can be retrieved from the proxy instead of upstream. 

Expose `CODER_PROVISIONER_TERRAFORM_{INIT,PLAN,APPLY}_ENV` on both `coderd` and `provisionerd` that will allow passing an array of environment variables into `terraform.provisionEnv`.

**NOTE:** It should **not** allow overriding any environment variable starting with the prefix `CODER_`.

**Comment by dannykopping** [comment]
> Expose `CODER_PROVISIONER_TERRAFORM_{INIT,PLAN,APPLY}_ENV` on both `coderd` and `provisionerd` that will allow passing an array of environment variables into `terraform.provisionEnv`.

Did you mean to add the `_0` suffix on `CODER_PROVISIONER_TERRAFORM_INIT_ENV_0`?

I think you should probably also specify how these env vars should be defined. Comma-separated values?

> NOTE: It should not allow overriding any environment variable starting with the prefix CODER_.

Good thinking!

**Comment by johnstcn** [comment]
> Did you mean to add the _0 suffix on CODER_PROVISIONER_TERRAFORM_INIT_ENV_0?

Yes, this is how `serpent.StringArrayOf()` works. Search for `CODER_EXTERNAL_AUTH_0.*` for more examples.
